### PR TITLE
Enable CPI Release golang

### PR DESCRIPTION
**Pull request by OBoehrer**
### Outcome of this PR

This PR is a follow up on https://github.com/cloudfoundry/bosh-openstack-cpi-release/pull/278, here we provide the complete implementation of the Golang-implemented job 'openstack_cpi_golang'. This job is completely independent, and can be used as an alternative to the job 'openstack_cpi'. 

### Background

Maintaining the existing Ruby CPI implies sustaining the used [openstack-fog](https://github.com/fog/fog-openstack) and [fog core](https://github.com/fog/fog) libraries. But that would be very problematic since the communities for these libraries do not seem to be active since some years now.

Fog isn't just a thin wrapper around the OpenStack API. On the contrary, Fog returns rich modules and classes that need to be refreshed, updated, ... . That means, replacing Fog calls by REST calls isn't a trivial task and will finally end up in a more or less complete re-implementation of the CPI.

A Golang based CPI reimplementation would profit from the existence of an actively maintained Golang SDK for OpenStack (see: https://github.com/gophercloud/gophercloud)

### Difference

This is not a 1:1 implementation of the Ruby CPI. We try to improve and change the structure as we best see fit. Deprecated APIs are not supported by this CPI, hence we only support the Octavia APIs, Compute APIs, Networking APIs ... and not the neutron/networking endpoints.

### Note

This golang implementation is based on the Ruby version as of Mar 22, 2024.

### CODE REVIEW

**Comment by fmoehler** [comment]
Co-authored by @a-hassanin, @anshrupani, @mvach,

**Comment by ShilpaChandrashekara** [comment]
> Co-authored by @a-hassanin, @anshrupani, @mvach, @ShilpaChandrashekara

File: jobs/openstack_cpi_golang/spec [hunk 1]
```diff
@@ -4,13 +4,10 @@ templates:
   cpi.erb: bin/cpi
   cpi.json.erb: config/cpi.json
   cacert.pem.erb: config/cacert.pem
-  cpi-golang.erb: bin/cpi-golang
-  cpi-ruby.erb: bin/cpi-ruby
 
 packages:
 - openstack_cpi_golang
 - bosh_openstack_cpi
```
Comment (aramprice) [inline jobs/openstack_cpi_golang/spec:10]: Is the ruby CPI source needed here?

Comment (a-hassanin) [reply jobs/openstack_cpi_golang/spec:10]: Not anymore, will remove it.

**Comment by a-hassanin** [comment]
These calls seem to be incompatible with Windows:

https://github.com/OBoehrer/bosh-openstack-cpi-release/blob/c54a6b48f5ec00b787ae7b7ab45894ff5820908a/src/openstack_cpi_golang/cpi/image/root_image/root_image_test.go#L15

https://github.com/OBoehrer/bosh-openstack-cpi-release/blob/c54a6b48f5ec00b787ae7b7ab45894ff5820908a/src/openstack_cpi_golang/cpi/methods/create_stemcell.go#L65

See Go / Run tests and Lint on windows-2019

**Review by a-hassanin — CHANGES_REQUESTED** [review]
These calls seem to be incompatible with Windows:

https://github.com/OBoehrer/bosh-openstack-cpi-release/blob/c54a6b48f5ec00b787ae7b7ab45894ff5820908a/src/openstack_cpi_golang/cpi/image/root_image/root_image_test.go#L15

https://github.com/OBoehrer/bosh-openstack-cpi-release/blob/c54a6b48f5ec00b787ae7b7ab45894ff5820908a/src/openstack_cpi_golang/cpi/methods/create_stemcell.go#L65

See Go / Run tests and Lint on windows-2019

File: src/openstack_cpi_golang/cpi/volume/volume_service_test.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
+package volume_test
```
Comment (a-hassanin) [inline src/openstack_cpi_golang/cpi/volume/volume_service_test.go:1]: Tests in this files are flaky, to reproduce the issue:

go clean -testcache 

and re-run the tests and repeat several times and sometimes they fail with no reason.

**Review by a-hassanin — CHANGES_REQUESTED** [review]

File: src/openstack_cpi_golang/cpi/volume/volume_service_test.go (hunk 2 above)
Comment (OBoehrer) [reply src/openstack_cpi_golang/cpi/volume/volume_service_test.go:1]: Reason for behavior: Race condition between timeout setting and API call.
Timeout handling in volume/snapshot get:
initial: 1 ns, leading sometimes to a timeout handling where a failing API call was expected
new: 1 minute, timeout cannot occur any more before failing API call

**Review by a-hassanin — APPROVED** [review]
LGTM

**Comment by aramprice** [comment]
Coverage for golang CPI code seems low:
```
 go test -cover ./...
?   	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/compute	[no test files]
?   	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/image	[no test files]
?   	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/loadbalancer	[no test files]
?   	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/network	[no test files]
?   	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/image/root_image	[no test files]
?   	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/openstack	[no test files]
?   	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/volume	[no test files]
	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang		coverage: 0.0% of statements
	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi		coverage: 0.0% of statements
ok  	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/config	(cached)	coverage: 97.4% of statements
ok  	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/methods	(cached)	coverage: 5.0% of statements
	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/mocks		coverage: 0.0% of statements
	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/utils		coverage: 0.0% of statements
ok  	github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/integration	(cached)	coverage: [no statements]
```

For an HTML version - see attached [coverage.html.zip](https://github.com/user-attachments/files/17511571/coverage.html.zip):
```

go test -coverprofile=coverage.out ./...
go tool cover -html=coverage.out -o coverage.hml
```

**Comment by aramprice** [comment]
I'd recommend adding a `.golangci.yml`, specifically having a longer [`timeout:`](https://github.com/cloudfoundry/bosh-warden-cpi-release/blob/master/.golangci.yml#L3) can help with windows linting  and updating the `go.yml` workflow to something like:

```yaml
name: go
on:
  push:
    paths:
    - 'src/openstack_cpi_golang/**'
  pull_request:
    paths:
    - 'src/openstack_cpi_golang/**'
jobs:
  lint_and_test:
    strategy:
      matrix:
        os: [macos-latest, ubuntu-latest]
    name: lint and test
    runs-on: ${{ matrix.os }}
    steps:
    - uses: actions/checkout@v4
    - uses: actions/setup-go@v5
      with:
        go-version-file: src/openstack_cpi_golang/go.mod
    - uses: golangci/golangci-lint-action@v6
      with:
        working-directory: src/openstack_cpi_golang/
    - run: run-unit-tests
      working-directory: src/openstack_cpi_golang/
    - run: run-integration-tests
      working-directory: src/openstack_cpi_golang/
```

modeled after [cloudfoundry/bosh-warden-cpi-release](https://github.com/cloudfoundry/bosh-warden-cpi-release)'s [`go.yml`](https://github.com/cloudfoundry/bosh-warden-cpi-release/blob/master/.github/workflows/go.yml) and adding a [`dependabot.yml`](https://github.com/cloudfoundry/bosh-warden-cpi-release/blob/master/.github/dependabot.yml) to keep Github Actions up to date.

**Review by aramprice — APPROVED** [review]

**Pull request merged**